onnxruntime/core/providers/dml/dml_provider_factory.cc

```diff
@@ -216,6 +265,128 @@
+        (ComPtr<IDXCoreAdapter>*) malloc(count * sizeof(ComPtr<IDXCoreAdapter>));
+    int npu_adapter_count = 0;
+
+    // Iterate through all compute capable adapters
+    for (uint32_t i = 0; i < count; ++i)
+    {
+        ComPtr<IDXCoreAdapter> candidateAdapter;
+        ORT_THROW_IF_FAILED(
+            d3D12CoreComputeAdapters->GetAdapter(i, candidateAdapter.GetAddressOf()));
+
+        // Only considering hardware adapters
+        if (!isHardwareAdapter(candidateAdapter))
+            continue;
+
+        if (f == OrtDmlDeviceFilter::Gpu) // consider GPUs only
+        {
+            if (supportsGraphics(candidateAdapter)) {
+                ordered_adapters[ordered_adapter_count] = candidateAdapter;
+                ordered_adapter_count++;
+            }
+        }
+        else if(f == OrtDmlDeviceFilter::Npu) // consider NPUs only
+        {
+            if (!supportsGraphics(candidateAdapter)) {
+                ordered_adapters[ordered_adapter_count] = candidateAdapter;
+                ordered_adapter_count++;
+            }
+        }
+        else // consider both GPUs and NPUs
+        {
+            if (supportsGraphics(candidateAdapter)) {
+                gpu_adapters[gpu_adapter_count] = candidateAdapter;
+                gpu_adapter_count++;
+            } else {
+                npu_adapters[npu_adapter_count] = candidateAdapter;
+                npu_adapter_count++;
+            }
+        }
+    }
+
+    if (f == OrtDmlDeviceFilter::None) // considering both NPUs and GPUs
+    {
+        // order the adapters based on the performance preference
+        ComPtr<IDXCoreAdapter>* high_pri_adapters;
+        ComPtr<IDXCoreAdapter>* low_pri_adapters;
+        int num_high_pri_adapters;
+        int num_low_pri_adapters;
+        if (p == OrtDmlPerformancePreference::LowPower) // NPUs first
+        {
+            high_pri_adapters = npu_adapters;
+            num_high_pri_adapters = npu_adapter_count;
+            low_pri_adapters = gpu_adapters;
+            num_low_pri_adapters = gpu_adapter_count;
+        }
+        else // GPUs first
+        {
+            high_pri_adapters = gpu_adapters;
+            num_high_pri_adapters = gpu_adapter_count;
+            low_pri_adapters = npu_adapters;
+            num_low_pri_adapters = npu_adapter_count;
+        }
+
+        for (int i = 0; i < num_high_pri_adapters; i++) {
+            ordered_adapters[ordered_adapter_count] = high_pri_adapters[i];
+            ordered_adapter_count++;
+        }
+        for (int i = 0; i < num_low_pri_adapters; i++) {
+            ordered_adapters[ordered_adapter_count] = low_pri_adapters[i];
+            ordered_adapter_count++;
+        }
+    }
+
+    // check if ordered_adapter count is 0 (no adapters?)
+    ComPtr<IDXCoreAdapter> highest_pri_adapter = ordered_adapters[0];
+    // return the create function for a dxcore device
+    options->provider_factories.push_back(onnxruntime::DMLProviderFactoryCreator::CreateDXCore(
```

Comment:
>CreateDXCore [](http://example.com/codeflow?start=82&length=12)

should this API take a list of adapters?

Reply:
By the time this API is reached the preferred device is already decided. Are you suggesting that this API should take the list of devices ordered by preference and try creating a DML device from each one in case of failure?

Reply:
In the future, DML may need to select over multiple devices, passing in the entire list in ordered form will be more robust going into the future.